config/config_params.go

```diff
@@ -404,27 +404,44 @@ func (config *Config) DatastoreConfig() apiconfig.CalicoAPIConfig {
 	}
 
 	// Now allow FELIX_XXXYYY variables or XxxYyy config file settings to override that, in the
-	// etcd case.
-	if config.setByConfigFileOrEnvironment("DatastoreType") && config.DatastoreType == "etcdv3" {
-		cfg.Spec.DatastoreType = apiconfig.EtcdV3
-		// Endpoints.
-		if config.setByConfigFileOrEnvironment("EtcdEndpoints") && len(config.EtcdEndpoints) > 0 {
-			cfg.Spec.EtcdEndpoints = strings.Join(config.EtcdEndpoints, ",")
-		} else if config.setByConfigFileOrEnvironment("EtcdAddr") {
-			cfg.Spec.EtcdEndpoints = config.EtcdScheme + "://" + config.EtcdAddr
-		}
-		// TLS.
-		if config.setByConfigFileOrEnvironment("EtcdKeyFile") {
-			cfg.Spec.EtcdKeyFile = config.EtcdKeyFile
-		}
-		if config.setByConfigFileOrEnvironment("EtcdCertFile") {
-			cfg.Spec.EtcdCertFile = config.EtcdCertFile
-		}
-		if config.setByConfigFileOrEnvironment("EtcdCaFile") {
-			cfg.Spec.EtcdCACertFile = config.EtcdCaFile
+	// etcd case. Note that that etcd options are set even if the DatastoreType isn't etcdv3.
+	// This allows the user to rely the default DatastoreType being etcdv3 and still being able
+	// to configure the other etcdv3 options. As of the time of this code change, the etcd options
+	// have no affect if the DatastoreType is not etcdv3.
+
+	// Datastore type, either etcdv3 or kubernetes
+	if config.setByConfigFileOrEnvironment("DatastoreType") {
+		log.Infof("Overriding DatastoreType from felix config to %s", config.DatastoreType)
+		if config.DatastoreType == string(apiconfig.EtcdV3) {
+			cfg.Spec.DatastoreType = apiconfig.EtcdV3
+		} else if config.DatastoreType == string(apiconfig.Kubernetes) {
```

Comment:
This would be a change, but I think it probably makes sense.

- The change is: if someone has `FELIX_DATASTORETYPE=kubernetes`, that would previously have been ignored, and now it will take effect.

- The concern about making sense is that there's no way with `FELIX_...` or felix.cfg of specifying the k8s-apiserver connection.

So you'd need a combination of `FELIX_DATASTORETYPE=kubernetes` and other non-FELIX settings like `CALICO_KUBECONFIG=/etc/kubeconfig` or `KUBECONFIG=/etc/kubeconfig` or `K8S_...=...`.

And, instead of `FELIX_DATASTORETYPE=kubernetes`, you could equally specify `DATASTORE_TYPE=kubernetes` or `CALICO_DATASTORE_TYPE=kubernetes`.  So you might question why `FELIX_DATASTORETYPE=kubernetes` needs to be allowed as well?

But I think the point is that someone might already have all the (non-CALICO-prefixed) `K8S....` or `KUBE...` variables set up in their environment, and then they should only need one thing to make Felix use those too, and then it's helpful if that can be any of `FELIX_DATASTORETYPE=kubernetes` or `DATASTORE_TYPE=kubernetes` or `CALICO_DATASTORE_TYPE=kubernetes`.  So, on balance, I think this is a helpful change.  WDYT?

Reply:
I definitely understand the concern, but I agree that this is a helpful change. The main reason I wanted this in here though was to keep the felix config overwriting simple. Maybe somebody will break their felix by specifying kubernetes in the felix configuration, but they should know that that will take affect, as opposed to the flip side where somebody specifies the other kubernetes env variables through the calico common configuration and specifies kubernetes in the felix configuration for the datastore (maybe in the scenerio you outlined) and they have no idea why the datastore is still etcd